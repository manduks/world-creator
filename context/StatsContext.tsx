import * as React from "react";
import { Coordinates } from 'types/world';

type StatsProviderProps = {children: React.ReactNode}
interface StatsInterface {
  cursor?: Coordinates;
  squaresCount?: number;
  islandsCount?: number;
}
interface Action extends StatsInterface { 
  type: 'cursor' | 'square' | 'reset'
}
type Dispatch = (action: Action) => void

const initialState = {
  cursor: {
    x: 0,
    y: 0,
  },
  squaresCount: 0,
  islandsCount: 0,
};
const StatsContext = React.createContext<StatsInterface | undefined>(undefined);
const StatsDispatchContext = React.createContext<Dispatch | undefined>(undefined);

function statsReducer(state: StatsInterface, action: Action) {
  switch (action.type) {
    case 'cursor': {
      return { ...state, cursor: action.cursor };
    }
    case 'square': {
      return { ...state, squaresCount: action.squaresCount, islandsCount:  action.islandsCount};
    }
    case 'reset': {
      return {
        ...initialState,
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

function StatsProvider({ children }: StatsProviderProps) {
  const [state, dispatch] = React.useReducer(statsReducer, initialState);

  return (
    <StatsContext.Provider value={state}>
      <StatsDispatchContext.Provider value={dispatch}>
        {children}
      </StatsDispatchContext.Provider>
    </StatsContext.Provider>
  )
}

function useStatsContext() {
  const context = React.useContext(StatsContext);
  if (context === undefined) {
    throw new Error('useStatsContext must be used within a StatsProvider')
  }
  return context;
}

function useStatsDispatch() {
  const context = React.useContext(StatsDispatchContext);
  if (context === undefined) {
    throw new Error('useStatsDispatch must be used within a StatsProvider')
  }
  return context;
}

export { StatsProvider, useStatsContext, useStatsDispatch };

