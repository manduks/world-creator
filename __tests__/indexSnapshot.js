import * as React from 'react'
import renderer from 'react-test-renderer'
import Index from '../pages/index'
import { StatsProvider } from 'context/StatsContext';

it('renders homepage unchanged', () => {
  const tree = renderer.create(
    <StatsProvider>
      <Index />
    </StatsProvider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})