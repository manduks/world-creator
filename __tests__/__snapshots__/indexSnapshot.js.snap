// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div
  className="container"
>
  <main>
    <div
      className="header"
    >
      <h1>
        Island Creator
      </h1>
      <span>
        Click on the canvas to start creating Islands
      </span>
    </div>
    <div
      className="world-container"
    >
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      >
        <div
          aria-label="grid"
          aria-readonly={true}
          className="ReactVirtualized__Grid"
          onScroll={[Function]}
          role="grid"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "ltr",
              "height": 0,
              "overflowX": "auto",
              "overflowY": "auto",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
    <footer>
      <div
        className="cursor"
      >
        <span>
          x : 
          0
        </span>
        <span>
          y : 
          0
        </span>
      </div>
      <div
        className="footer-inputs"
      >
        <div
          className="input-container"
        >
          <label
            title="Width:"
          >
            Width:
          </label>
          <input
            aria-label="Width:"
            max={100000}
            min={1}
            name="width"
            onChange={[Function]}
            placeholder="width"
            type="number"
            value={100}
          />
        </div>
        <div
          className="input-container"
        >
          <label
            title="Height:"
          >
            Height:
          </label>
          <input
            aria-label="Height:"
            max={100000}
            min={1}
            name="height"
            onChange={[Function]}
            placeholder="height"
            type="number"
            value={100}
          />
        </div>
      </div>
      <div
        className="stats"
      >
        <div
          data-testid="stats-squares"
        >
          <span
            className="square"
          />
           
          0
           squares
        </div>
        <div>
          <span
            data-testid="stats-islands"
          />
           
          0
           islands
        </div>
      </div>
    </footer>
  </main>
</div>
`;
